CRM/Core/BAO/Navigation.php

```diff
@@ -252,7 +252,7 @@ public static function buildNavigationTree() {
 
     // get the list of menus
     $query = "
-SELECT id, label, url, permission, permission_operator, has_separator, parent_id, is_active, name
+SELECT id, label, url, permission, permission_operator, has_separator, parent_id, is_active, name, icon 
```

Comment:
@omarabuhussein i feel like this and below should be made conditional if the field is present or not just in case if there are pre-upgrade issues

Reply:
@omarabuhussein this is the sort of change i am thinking of

```
diff --git a/CRM/Core/BAO/Navigation.php b/CRM/Core/BAO/Navigation.php
index b7c5a01491..7a738616e6 100644
--- a/CRM/Core/BAO/Navigation.php
+++ b/CRM/Core/BAO/Navigation.php
@@ -251,8 +251,12 @@ FROM civicrm_navigation WHERE domain_id = $domainID {$whereClause} ORDER BY pare
     $navigationTree = array();

     // get the list of menus
+    $fields = "SELECT id, label, url, permission, permission_operator, has_separator, parent_id, is_active, name";
+    if (!CRM_Core_Config::isUpgradeMode() && CRM_Core_DAO::checkFieldExists('civicrm_navigation', 'icon', FALSE)) {
+      $fields .= ",icon";
+    }
     $query = "
-SELECT id, label, url, permission, permission_operator, has_separator, parent_id, is_active, name, icon
+{$fields}
 FROM civicrm_navigation
 WHERE domain_id = $domainID
 ORDER BY parent_id, weight";
@@ -264,7 +268,6 @@ ORDER BY parent_id, weight";
           'label' => $navigation->label,
           'name' => $navigation->name,
           'url' => $navigation->url,
-          'icon' => $navigation->icon,
           'permission' => $navigation->permission,
           'operator' => $navigation->permission_operator,
           'separator' => $navigation->has_separator,
@@ -273,6 +276,9 @@ ORDER BY parent_id, weight";
           'active' => $navigation->is_active,
         ),
       );
+      if (!CRM_Core_Config::isUpgradeMode() && CRM_Core_DAO::checkFieldExists('civicrm_navigation', 'icon', FALSE)) {
+        $navigationTree[$navigation->id]['attributes']['icon'] = $navigation->icon;
+      }
     }

     return self::buildTree($navigationTree);
```

Reply:
thanks for the review @seamuslee001 , but following this logic we need to fill the code with tests for availability of any column before using it (each time new column introduced , more checks to be added) and eventually the code will get polluted and complicated with time.

I believe checking the integrity of upgrades should be done via external tool (or maybe adding new admin page for it) and not be checked for inside the code itself.

Reply:
@omarabuhussein The difference here is that it can be pretty noticeable straight forward upon applying this code as it was on my local buildkit install i lost the admin menu because it was having a db error. There are only certain times i think its right to do this such as looking at this https://github.com/civicrm/civicrm-core/pull/10974 or whereby `CRM_Core_BAO_CustomGroup::getTree` is called in building cache so that could bork if its trying to read off fields that don't exist in the db yet

Reply:
I think this indicates that the upgrader code need to be reviewed and fixed, I mean why an upgrader code would even try to rebuild the navigation menu which is not relevant to the upgrade process.

But I took this a step further and wanted to try it locally by following these steps : 

1- Creating new civicrm 4.7.22 local site.
2- Replacing civicrm codebase with the latest available version which is 4.7.44, I downloaded it from this link : https://download.civicrm.org/civicrm-4.7.24-drupal.tar.gz .
3- changing CRM_Core_BAO_Navigation, CRM_Core_DAO_Navigation and CRM_Upgrade_Incremental_php_FourSeven to match the files in my PR (in other words, applying the relevant part in this PR to 4.7.24).
4- I checked civi at this point, I also tried to edit the menu and reset the cache but the menu was still there. I also used xdebug just to make sure everything is fine and  traceed this line (https://github.com/omarabuhussein/civicrm-core/blob/1a0841828512835432a4629460ef66928cc60a0a/CRM/Core/BAO/Navigation.php#L264)
and the icon was set to null and the code did not trigger any error. Are you sure the db error you had was relevant to navigation menu ? also despite that I don't think it is important for Civi to usable while it is still in (need upgrade) process so even if the menu was missing then that's fine, it means the user need to run the db upgrader.
5- I continued the test by running the upgrader and it went fine without a problem and the icon field was created successfully.

I did all of that because I really  didn't want to add such check to the code, and while it could be necessary at some cases, my test above suggests that it is not in this case. So what do you think ?